File: golang/src/jomiel.github.io/demo/examples/jomiel.go
/*
 * -*- coding: utf-8 -*-
 *
 * jomiel-examples
 *
 * Copyright
 *  2019 Toni Gündoğdu
 *
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package demo

import (
	"encoding/json"
	"fmt"
	"log"

	proto "github.com/golang/protobuf/proto"
	czmq "gopkg.in/zeromq/goczmq.v4"

	pb "jomiel.github.io/demo/examples/proto"
)

type Jomiel struct {
	sock *czmq.Sock
	opts *Options
}

func NewJomiel(opts *Options) *Jomiel {
	sck := czmq.NewSock(czmq.Req)
    sck.SetOption(czmq.SockSetLinger(0))
	return &Jomiel{
		opts: opts,
		sock: sck,
	}
}

func (j *Jomiel) Destroy() {
	defer j.sock.Destroy()
}

func (j *Jomiel) Connect() {
	re := j.opts.RouterEndpoint
	to := j.opts.ConnectTimeout

	status := fmt.Sprintf("<connect> %s (timeout=%d)", re, to)

	j.printStatus(status)
	j.sock.Connect(j.opts.RouterEndpoint)
}

func (j *Jomiel) Inquire(uri string) {
	j.send(uri)
	j.recv()
}

func (j *Jomiel) send(uri string) {
	inquiry := &pb.Inquiry{
		Inquiry: &pb.Inquiry_Media{
			Media: &pb.MediaInquiry{
				InputUri: uri,
			},
		},
	}

	out, err := proto.Marshal(inquiry)
	if err != nil {
		log.Fatalln("failed to encode inquiry: ", err)
	}

	j.printStatus("<send>")
	j.sock.SendMessage([][]byte{out})
}

func (j *Jomiel) recv() {
	poller, err := czmq.NewPoller(j.sock)
	if err != nil {
		log.Fatalln("failed to create a poller: ", err)
	}
	defer poller.Destroy()

	sck, err := poller.Wait(j.opts.ConnectTimeout * 1000)
	if err != nil {
        log.Fatalln("failed to pollin an event: ", err)
	} else {
        if sck == nil {
            log.Fatalln("error: connection timed out")
        }
    }

	data, err := j.sock.RecvMessage()
	if err != nil {
		log.Fatalln("failed to receive message: ", err)
	}

	response := &pb.Response{}
	err = proto.Unmarshal(data[0], response)
	if err != nil {
		log.Fatalln("failed to decode response: ", err)
	}

	j.dumpResponse(response)
}

func (j *Jomiel) printStatus(status string) {
	if !j.opts.BeTerse {
		log.Println("status: " + status)
	}
}

func (j *Jomiel) toJson(response *pb.Response) string {
	json, err := json.MarshalIndent(response, "", "  ")
	if err != nil {
		log.Fatalln("error: ", err)
	}
	return string(json)
}

func (j *Jomiel) printMessage(status string, message *pb.Response) {
	var result string
	if j.opts.OutputJson {
		result = j.toJson(message)
	} else {
		result = proto.MarshalTextString(message)
	}
	j.printStatus(status)
	fmt.Printf(result)
}

func (j *Jomiel) dumpResponse(response *pb.Response) {
	responseStatus := response.GetStatus()
	mediaResponse := response.GetMedia()

	if responseStatus.GetCode() == pb.StatusCode_OK {
		if j.opts.BeTerse {
			j.dumpTerseResponse(mediaResponse)
		} else {
			j.printMessage("<recv>", response)
		}
	} else {
		j.printMessage("<recv>", response)
	}
}

func (j *Jomiel) dumpTerseResponse(response *pb.MediaResponse) {
	fmt.Println("---\ntitle: " + response.GetTitle())
	fmt.Println("quality:")

	stream := response.GetStream()

	for i := 0; i < len(stream); i++ {
		quality := stream[i].GetQuality()
		fmt.Printf("  profile: %s\n    width: %d\n    height: %d\n",
			quality.GetProfile(), quality.GetWidth(), quality.GetHeight())
	}
}

File: php7/lib/Jomiel.php
<?php
/* -*- coding: utf-8 -*-
 *
 * jomiel-examples
 *
 * Copyright
 *  2019 Toni Gündoğdu
 *
 *
 * SPDX-License-Identifier: Apache-2.0
 */
class Jomiel {
    private $ctx, $sck;
    private $opts, $lg;

    function __construct($logger, $opts) {
        $this->lg = $logger;
        $this->opts = $opts;

        $this->ctx = new ZMQContext;
        $this->sck = new ZMQSocket($this->ctx, ZMQ::SOCKET_REQ);

        $this->sck->setSockOpt(ZMQ::SOCKOPT_LINGER, 0);
    }

    function __destruct() {
        $this->sck->disconnect($this->opts->routerEndpoint);
    }

    function connect() {
        $addr = $this->opts->routerEndpoint;
        $time = $this->opts->connectTimeout;
        $status = sprintf("<connect> %s (timeout=%d)", $addr, $time);
        $this->printStatus($status);
        $this->sck->connect($addr);
    }

    function inquire($uri) {
        $this->send($uri);
        $this->recv();
    }

    private function send($uri) {
        $media_inquiry = new Jomiel\Media\MediaInquiry;
        $inquiry = new Jomiel\Inquiry;

        $media_inquiry->setInputUri($uri);
        $inquiry->setMedia($media_inquiry);

        if (! $this->opts->beTerse) {
            $this->printMessage("<send>", $inquiry);
        }

        $bytes = $inquiry->serializeToString();
        $this->sck->send($bytes);
    }

    private function recv() {
        $poll = new ZMQPoll;
        $poll->add($this->sck, ZMQ::POLL_IN);

        $timeout = $this->opts->connectTimeout * 1000; # to msec

        $writeable = array();
        $readable = array();
        $events = 0;

        try {
            $events = $poll->poll($readable, $writeable, $timeout);
            $errors = $poll->getLastErrors();

            if (count($errors) >0) {
                foreach($errors as $error) {
                    $this->lg->error($error);
                }
                die(1);
            }
        } catch (ZMQPollException $e) {
            $this->lg->error($e->getMessage());
            die(1);
        }

        if ($events >0) {
            foreach ($readable as $r) {
                try {
                    $bytes = $this->sck->recv();
                    $response = new Jomiel\Response;
                    $response->mergeFromString($bytes);
                    $this->dumpResponse($response);
                } catch (Exception $e) {
                    $this->lg->error($e->getMessage());
                    die(1);
                }
            }
        } else {
            $this->lg->error("connection timed out");
            die(1);
        }
    }

    private function getQualityString($stream_quality) {
        return sprintf("  profile: %s\n    width: %d\n    height: %d",
                        $stream_quality->getProfile(),
                        $stream_quality->getWidth(),
                        $stream_quality->getHeight());
    }

    private function dumpTerseResponse($media_response) {
        echo "---\ntitle: " . $media_response->getTitle() ."\n";
        echo "quality:\n";
        foreach ($media_response->getStream() as $stream) {
            $stream_quality = $stream->getQuality();
            $quality_string = $this->getQualityString($stream_quality);
            echo "$quality_string\n";
        }
    }

    private function dumpResponse($response) {

        $status = $response->getStatus();

        if ($status->getCode() == Jomiel\Status\StatusCode::OK) {
            $media_response = $response->getMedia();
            if ($this->opts->beTerse) {
                $this->dumpTerseResponse($media_response);
            } else {
                $this->printMessage("<recv>", $media_response);
            }
        } else {
            $this->printMessage("<recv>", $response);
        }
    }

    private function printMessage($status, $message) {
        $this->printStatus($status);
        echo $message->serializeToJsonString() . "\n";
    }

    private function printStatus($status) {
        if (! $this->opts->beTerse) {
            $this->lg->info($status);
        }
    }
}
?>
